Makefile

```diff
@@ -51,6 +51,9 @@ libtree-sitter.$(SOEXTVER): $(OBJ)
 	ln -sf $@ libtree-sitter.$(SOEXT)
 	ln -sf $@ libtree-sitter.$(SOEXTVER_MAJOR)
 
+%.o: %.c
+	$(CC) $(CFLAGS) -c $^ -o $@
```

Comment:
If you're going to re-implement implicit rules, you should at least implement them correctly. This should include `CPPFLAGS`.

That said, implicit rules are one of make's greatest strengths, and what allows even big projects to have somewhat concise Makefiles. As a packager, I greatly dislike dealing with Makefiles that try to re-implement make functionality and get it wrong or hit some corner case, when the implicit rules would have worked just fine.

A makefile is written expecting that it will be used by a user calling `make`, with the addition of their own compiler/linker flags or compiler. If you disable functionality that is expected of any Make utility (it's even in the POSIX spec), you're mostly on your own. A project that requires their makefile to be run with `make -r` should specify that, not the contrary.